### ISSUE TITLE
Compiler crashes when check expressions are used in from queries when the error would be caught by an on fail clause 

### ISSUE THREAD

**Issue by buehlefs** [ballerina-platform/ballerina-lang#34894]
**Description:**
A from query containing a check expression can trip up the compiler when the error is not directly returned by the surrounding function but caught by an `on fail` clause of a `do` or a `transaction` block.

**Steps to reproduce:**

Try to compile the minimal example below:

```ballerina
import ballerina/io;
public function main() {
    do {
        _ = from string v in []
            select check someFunction(v);
    } on fail var err {
        io:println(err);
    }
}

function someFunction(string input) returns string|error {
    return input;
}
```

<details>

<summary>
Click here to show more information.
</summary>

Stacktrace:
```text
java.lang.NullPointerException
        at org.wso2.ballerinalang.compiler.bir.codegen.JvmTerminatorGen.genGoToTerm(JvmTerminatorGen.java:386)
        at org.wso2.ballerinalang.compiler.bir.codegen.JvmTerminatorGen.genTerminator(JvmTerminatorGen.java:329)
        at org.wso2.ballerinalang.compiler.bir.codegen.methodgen.MethodGen.generateBasicBlocks(MethodGen.java:486)
        at org.wso2.ballerinalang.compiler.bir.codegen.methodgen.MethodGen.genJMethodForBFunc(MethodGen.java:235)
        at org.wso2.ballerinalang.compiler.bir.codegen.methodgen.MethodGen.generateMethod(MethodGen.java:161)
        at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.lambda$generateModuleClasses$0(JvmPackageGen.java:446)
        at java.base/java.util.HashMap$EntrySet.forEach(HashMap.java:1039)
        at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.generateModuleClasses(JvmPackageGen.java:397)
        at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.generate(JvmPackageGen.java:811)
        at org.wso2.ballerinalang.compiler.bir.codegen.CodeGenerator.generate(CodeGenerator.java:95)
        at org.wso2.ballerinalang.compiler.bir.codegen.CodeGenerator.generate(CodeGenerator.java:70)
        at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:301)
        at io.ballerina.projects.ModuleContext.generateCodeInternal(ModuleContext.java:466)
        at io.ballerina.projects.ModuleCompilationState$4.generatePlatformSpecificCode(ModuleCompilationState.java:132)
        at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:389)
        at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:159)
        at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:137)
        at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:113)
        at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
        at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:182)
        at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:112)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:93)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.BuildCommand.execute(BuildCommand.java:249)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.ballerina.cli.launcher.Main.main(Main.java:51)
```

The following code fails with a different stacktrace:

```ballerina
import ballerina/io;
public function main() {
    transaction { // this leads to a different stacktrace
        _ = check from string v in []
            do { _ = check someFunction(v); }; // just another way to trigger the bug?
        check commit;
    } on fail var err {
        io:println(err);
    }
}

function someFunction(string input) returns string|error {
    return input;
}
```

Stacktrace:
```text
java.lang.NullPointerException
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:164)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRBasicBlock.accept(BIRNode.java:443)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.lambda$visit$4(BIROptimizer.java:126)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:126)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRFunction.accept(BIRNode.java:401)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.lambda$visit$1(BIROptimizer.java:103)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:103)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRPackage.accept(BIRNode.java:83)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer.optimizePackage(BIROptimizer.java:82)
        at org.wso2.ballerinalang.compiler.bir.BIRGen.genBIR(BIRGen.java:269)
        at io.ballerina.projects.internal.CompilerPhaseRunner.birGen(CompilerPhaseRunner.java:220)
        at io.ballerina.projects.internal.CompilerPhaseRunner.performBirGenPhases(CompilerPhaseRunner.java:157)
        at io.ballerina.projects.ModuleContext.generateCodeInternal(ModuleContext.java:452)
        at io.ballerina.projects.ModuleCompilationState$4.generatePlatformSpecificCode(ModuleCompilationState.java:132)
        at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:389)
        at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:159)
        at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:137)
        at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:113)
        at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
        at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:182)
        at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:112)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:93)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.BuildCommand.execute(BuildCommand.java:249)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.ballerina.cli.launcher.Main.main(Main.java:51)
```

The `transaction` block causes the error even when the `on fail` clause is missing:

```ballerina
public function main() returns error? {
    transaction {
        _ = check from string v in []
            do {_ = check someFunction(v);};
    } 
}
```

But changing it to a `do` block will compile just fine.

</details>

Possible Workaround until the issue is resolved:

```ballerina
// trap panic, convert back to error
_ = check trap from string v in []
    select checkpanic someFunction(v); // panic on error
```

**Affected Versions:**

Tested on `slbeta6` but the error probably exists since `slbeta4`. Similar code has compiled fine with `slbeta3`.

**OS, DB, other environment details and versions:**

Ubuntu (on WSL2)

**Related Issues:**

Probably related: #34557 

There are more issues with similar stacktraces of compiler errors.

**Suggested Labels:**

Team/CompilerFE, Area/BIR?

**Comment by pcnfernando** [comment]
As per the observation error is not returned from the `select` clause. 

```ballerina
public function main() returns error? {
    do {
        var res = from string v in []
            select check someFunction(v);
        io:println(res);
    } 
}

function someFunction(string input) returns string|error {
    return error("dummy error");
}
```

**Comment by lochana-chathura** [comment]
Fixed in #35344

**Comment by pcnfernando** [comment]
The original issue is not fixed with #35344 but only https://github.com/ballerina-platform/ballerina-lang/issues/34894#issuecomment-1028744801. Hence; reopening the issue.

**Comment by pcnfernando** [comment]
The fix is available with 2201.1.0
